benchmark-automation/requirements.md

```diff
@@ -0,0 +1,125 @@
+
+## Out of scope
+
+We aren't aiming to support an "implementation shootout", e.g. between PyPy, Pyston and CPython.
+The pyperformance benchmark suite will continue to be designed for this, but the rest of the tooling described here may not necessarily support it.
+
+## Block diagram
+
+The following is a block diagram of the proposed system.
+
+While this is not intended as a description of current functionality, existing tools that provide some of this functionality are noted.
+Importantly, there isn't currently a connection between the benchmark runner and `speed.python.org` -- these are currently distinct efforts.
+
+![Block diagram](/benchmark-automation/Benchmark Infrastructure Block Diagram.png)
+
+## Definitions
+
+### Results
+
+"Results", all of which come from running `pyperformance` in the same configuration, come in the following flavors:
+
+- **Nightly**: Run once a day on the current `main` branch of CPython.
+- **Tag**: Run on tagged releases of Python. The most recent patch release for all supported releases, and all alphas, betas and rcs for the current development release.
+- **Manual**: Run on a specific commit or pull request of CPython. Usually created in response to Github automation or using a CLI tool.
+
+## Requirements
+
+### Requesting a manual benchmark run at the commandline
+
+Using a commandline tool available on PyPI, a developer could request a benchmark run on a specific commit in their published fork of CPython.
+The results would be uploaded to the results database and viewable on the public website, with an option to download the complete raw data.
+
+For example:
+
+```
+$ run_benchmark myname/cpython@135abf
+Results will be available at https://example.com/pyspeed/myname/135abf
+```
+
+For security and scaling reasons, users would need to be granted explicit permission to do this. There may be a way to piggy back on Github authentication (which would let permitted users be managed there), but TBD whether this is possible.
+
+### Requesting a manual benchmark in a pull request
+
+When a contributor submits a pull request where a performance impact is expected, a developer with sufficient permissions could request a benchmark run.
+This could be done with, for example, optional Github Actions or a bot that monitors pull requests.
+
+### Benchmark suite evolution
+
+We expect the benchmark suite to evolve rapidly in the near future as we add more real-world workloads to it.
+When a new release of `pyperformance` ships, the following will happen automatically:
+
+- Rerun all **Tag** results.
+- Rerun the last two weeks of **Nightly** results.
+- **Manual** results will be invalidated, and will need to be re-requested against the new benchmark suite.
+
+To avoid problems, the front-end analysis tools will be designed to never allow comparing results from two different releases of `pyperformance`.
```

Comment:
I wonder if we could at least compare results of analogous runs (i.e., from the same git revision) for two adjacent PyPerformance releases, so we can assess whether the new PyPerformance release changes anything? And then we might (manually, after some deliberation) declare the new version compatible with the old version.

Reply:
Yeah, that probably makes some sense.  In the general case, all bets are off, but if we know and can confirm manually it has no impact, we could flag something like "this tagged release is compatible with the previous one".